### PR TITLE
Make plugin

### PR DESCRIPTION
**Pull request by adler-j**
Work in progress.

Changes makes this a plugin instead of a subpackage.

### CODE REVIEW

File: odlpp/CMakeLists.txt [hunk 1]
```diff
@@ -18,12 +18,21 @@ set(INSTALL_LIB "pyinstall")
 add_custom_target(${INSTALL_LIB} COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/__init__.py ${TARGET_PY}/__init__.py)
 set_property(TARGET ${INSTALL_LIB} PROPERTY FOLDER python)
 
+# copy plugin manager
+add_custom_command(TARGET ${INSTALL_LIB} COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/odl_pluggin.py ${TARGET_PY}/odl_pluggin.py)
```
Comment (kohr-h) [inline odlpp/CMakeLists.txt]: `odl_plugin.py`?

File: odlpp/CMakeLists.txt [hunk 2]
```diff
@@ -18,12 +18,21 @@ set(INSTALL_LIB "pyinstall")
...
+add_custom_command(TARGET ${INSTALL_LIB} COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/setup.py ${TARGET_PY}/setup.py)
+add_custom_command(TARGET ${INSTALL_LIB} COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/cu_ntuples.py ${TARGET_PY}/cu_ntuples.py)
+add_custom_command(TARGET ${INSTALL_LIB} COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/ufuncs.py ${TARGET_PY}/ufuncs.py)
+MESSAGE( STATUS "GOT TO COPYING" )
```
Comment (kohr-h) [inline odlpp/CMakeLists.txt]: Temporary I guess

File: odlpp/cu_ntuples.py [hunk 3]
```diff
@@ -0,0 +1,1458 @@
...
+from odl.util.utility import dtype_repr
+
+from odlpp.ufuncs import CudaNtuplesUFuncs
+
+try:
+    import odlpp.odlpp_cuda as cuda
+    CUDA_AVAILABLE = True
+except ImportError:
+    cuda = None
+    CUDA_AVAILABLE = False
```
Comment (kohr-h) [inline odlpp/cu_ntuples.py]: Any more need for this? Or do we keep it for consistency?

File: odlpp/cu_ntuples.py [hunk 4]
```diff
@@ -0,0 +1,1458 @@
...
+        """
+        if inp is None:
+            if data_ptr is None:
+                return self.element_type(self, self._vector_impl(self.size))
+            else:  # TODO: handle non-1 length strides
+                return self.element_type(
+                    self, self._vector_impl.from_pointer(data_ptr, self.size,
+                                                         1))
+        else:
+            if data_ptr is None:
```
Comment (kohr-h) [inline odlpp/cu_ntuples.py]: I think we should write this as

``` python
if inp is None and data_ptr is None:
    ...
elif inp is None and data_ptr is not None:
    ...
elif inp is not None and data_ptr is None:
    ...
else:
    ...
```

That makes the logic much less cryptic.

File: odlpp/cu_ntuples.py [hunk 5]
```diff
@@ -0,0 +1,1458 @@
...
+                        elem[:] = inp
+                        return elem
+                else:
+                    # Array-like input. Need to go through a NumPy array
+                    arr = np.array(inp, copy=False, dtype=self.dtype, ndmin=1)
+                    if arr.shape != (self.size,):
+                        raise ValueError('expected input shape {}, got {}.'
+                                         ''.format((self.size,), arr.shape))
+                    elem = self.element()
+                    elem[:] = arr
```
Comment (kohr-h) [inline odlpp/cu_ntuples.py]: ~~We still have different behavior for CPU and CUDA Ntuples: the CPU version [does not allow](https://github.com/odlgroup/odl/blob/94c4b28f676674bb7f4561d1c0832a49d79171d6/odl/space/numpy_ntuples.py#L121-L125) `inp` to be a single value (unless `size == 1`) while this code does. We should harmonize this. I'd say don't allow it.~~ Oops, missed that part above.

File: odlpp/cu_ntuples.py [hunk 6]
```diff
@@ -0,0 +1,1458 @@
...
+            Base class for ufuncs in `NtuplesBase` spaces.
+        """
+        return CudaNtuplesUFuncs(self)
+
+
+class CudaFn(FnBase, CudaNtuples):
+
+    """The space `FnBase`, implemented in CUDA.
+
+    Requires the compiled ODL extension odlpp.
```
Comment (kohr-h) [inline odlpp/cu_ntuples.py]: Can be removed.

File: odlpp/cu_ntuples.py [hunk 7]
```diff
@@ -0,0 +1,1458 @@
...
+_add_if_exists(np.int64, 'CudaVectorInt64')
+_add_if_exists(np.uint8, 'CudaVectorUInt8')
+_add_if_exists(np.uint16, 'CudaVectorUInt16')
+_add_if_exists(np.uint32, 'CudaVectorUInt32')
+_add_if_exists(np.uint64, 'CudaVectorUInt64')
+CUDA_DTYPES = list(set(CUDA_DTYPES))  # Remove duplicates
+
+
+class CudaNtuples(NtuplesBase):
+    """The space `NtuplesBase`, implemented in CUDA."""
```
Comment (kohr-h) [inline odlpp/cu_ntuples.py]: Blank row before

File: test/cudaspace_speedtest.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
+# Copyright 2014-2016 The ODL development group
```
Comment (kohr-h) [inline test/cudaspace_speedtest.py]: Do we want this in the test folder?

File: odlpp/CMakeLists.txt (hunk 1 above)
Comment (adler-j) [inline odlpp/CMakeLists.txt]: `odl_pluggggin.py`

**Comment by kohr-h** [comment]
Looking good up to minor things. I'll try to build and install myself first though.

File: odlpp/cu_ntuples.py [hunk 9]
```diff
@@ -0,0 +1,1458 @@
...
+    impl = 'cuda'
+
+    @staticmethod
+    def available_dtypes():
+        """Return the available data types."""
+        return CUDA_DTYPES.copy()
+
+    @staticmethod
+    def default_dtype(field=None):
+        """Return the default of `Fn` data type for a given field.
```
Comment (adler-j) [inline odlpp/cu_ntuples.py]: Remove `Fn`

File: odlpp/cu_ntuples.py [hunk 10]
```diff
@@ -0,0 +1,1458 @@
...
+                             ''.format(field))
+
+    def _lincomb(self, a, x1, b, x2, out):
+        """Linear combination of ``x1`` and ``x2``, assigned to ``out``.
+
+        Calculate ``z = a * x + b * y`` using optimized CUDA routines.
+
+        Parameters
+        ----------
+        a, b : `LinearSpace.field` `element`
```
Comment (adler-j) [inline odlpp/cu_ntuples.py]: All of these should be `odl.LinearSpace.field`, etc.

**Comment by kohr-h** [comment]
I didn't look too closely on the doc since I'm anyway doing a big overhaul. Design-wise it looks good.

**Comment by kohr-h** [comment]
Okay, after some playing around and observing that `conda` tries to be smart and not to copy stuff around more than it has to, I finally got a freshly built `odlpp` installed. When trying to import it, I get

```
File "/home/hkohr/miniconda3/envs/odlpp_test_27/lib/python2.7/site-packages/odlpp/__init__.py", line 23, in <module>
    from . import odlpp_cuda
ImportError: cannot import name odlpp_cuda
```

Looking at the output from the build process, it's no surprise:

```
...
[100%] Linking CXX shared library ../../bin/libodlpp_cuda.so
[100%] Built target odlpp_cuda
Scanning dependencies of target pyinstall
running install
running bdist_egg
running egg_info
creating odlpp.egg-info
writing requirements to odlpp.egg-info/requires.txt
writing odlpp.egg-info/PKG-INFO
writing top-level names to odlpp.egg-info/top_level.txt
writing dependency_links to odlpp.egg-info/dependency_links.txt
writing entry points to odlpp.egg-info/entry_points.txt
writing manifest file 'odlpp.egg-info/SOURCES.txt'
reading manifest file 'odlpp.egg-info/SOURCES.txt'
writing manifest file 'odlpp.egg-info/SOURCES.txt'
installing library code to build/bdist.linux-x86_64/egg
running install_lib
running build_py
creating build
creating build/lib
creating build/lib/odlpp
copying ./odl_pluggin.py -> build/lib/odlpp
copying ./cu_ntuples.py -> build/lib/odlpp
copying ./__init__.py -> build/lib/odlpp
copying ./ufuncs.py -> build/lib/odlpp
creating build/bdist.linux-x86_64
creating build/bdist.linux-x86_64/egg
creating build/bdist.linux-x86_64/egg/odlpp
copying build/lib/odlpp/odl_pluggin.py -> build/bdist.linux-x86_64/egg/odlpp
copying build/lib/odlpp/cu_ntuples.py -> build/bdist.linux-x86_64/egg/odlpp
copying build/lib/odlpp/__init__.py -> build/bdist.linux-x86_64/egg/odlpp
copying build/lib/odlpp/ufuncs.py -> build/bdist.linux-x86_64/egg/odlpp
byte-compiling build/bdist.linux-x86_64/egg/odlpp/odl_pluggin.py to odl_pluggin.pyc
byte-compiling build/bdist.linux-x86_64/egg/odlpp/cu_ntuples.py to cu_ntuples.pyc
byte-compiling build/bdist.linux-x86_64/egg/odlpp/__init__.py to __init__.pyc
byte-compiling build/bdist.linux-x86_64/egg/odlpp/ufuncs.py to ufuncs.pyc
creating build/bdist.linux-x86_64/egg/EGG-INFO
copying odlpp.egg-info/PKG-INFO -> build/bdist.linux-x86_64/egg/EGG-INFO
copying odlpp.egg-info/SOURCES.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying odlpp.egg-info/dependency_links.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying odlpp.egg-info/entry_points.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying odlpp.egg-info/requires.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying odlpp.egg-info/top_level.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
zip_safe flag not set; analyzing archive contents...
odlpp.__init__: module references __file__
creating dist
creating 'dist/odlpp-0.2.0-py2.7.egg' and adding 'build/bdist.linux-x86_64/egg' to it
removing 'build/bdist.linux-x86_64/egg' (and everything under it)
Processing odlpp-0.2.0-py2.7.egg
creating /home/hkohr/miniconda3/envs/_build/lib/python2.7/site-packages/odlpp-0.2.0-py2.7.egg
Extracting odlpp-0.2.0-py2.7.egg to /home/hkohr/miniconda3/envs/_build/lib/python2.7/site-packages
Adding odlpp 0.2.0 to easy-install.pth file

Installed /home/hkohr/miniconda3/envs/_build/lib/python2.7/site-packages/odlpp-0.2.0-py2.7.egg
```

The compiled stuff is never touched and does not end up in the final package. Thus, only a skeleton package is installed.
I'm not quite sure why it is like that. Possible explanations are (1) the CMake commands never copy files or (2) `conda` wraps up the package before the libraries are copied or ignores them in some other way.
I can build the code by hand to exclude reason (1). If it's related to (2), then probably we really need a `setup.py` based build so `conda` will see all the output, or some other way to manually copy the compiled files.

**Comment by kohr-h** [comment]
Same output for building "by hand", library files are not copied.

**Comment by adler-j** [comment]
Absolutely no idea why it would not do that, it does the copy on windows but I had some issues with it (see the `__path__` thing). Could you try checking the `package_data` and see if you can get it to work on linux?

**Comment by kohr-h** [comment]
If I replace the `package_data` in `odlpp/setup.py` by `package_data={'' : ['*.pyd', '*.so']}`, the files are included, and things finally work. Testing the conda build now.

**Comment by adler-j** [comment]
¯\_(ツ)_/¯

Great!

**Comment by kohr-h** [comment]
I think the proper way of doing this is to create a `MANIFEST.in` file (see [here](https://github.com/odlgroup/odl/blob/master/MANIFEST.in)) that contains all files apart from Python scripts/modules that are to be included in the distribution.

**Comment by kohr-h** [comment]
Added a [MANIFEST.in](https://github.com/odlgroup/odlpp/blob/71db46cd619793bf60d8aa625dbe528a36710841/MANIFEST.in) and it is actually used during the `setup.py install` command. However, it does not find the README.md and LICENSE files since they are not copied to the build directory. Here are the contents of `build/bin` where `python setup.py install` is run:

```
build          dist         libodlcpputils_cuda.a    libodlcpputils_test.a  MANIFEST.in     odlpp_cuda.so   setup.py
cu_ntuples.py  __init__.py  libodlcpputils_python.a  libodlpp_cuda.so       odl_pluggin.py  odlpp.egg-info  ufuncs.py
```

~~I know too little about CMake to figure out why it behaves like this. Perhaps the `.in` suffix makes it believe it's a config file or similar and copies it by default.~~ Stupid, I created the file in `build/bin` and copied it over. Some way to make sure that ~~also the other~~ all metadata files are copied would be good.

**Comment by kohr-h** [comment]
> Some way to make sure that also the other all metadata files are copied would be good.

Ping. Shouldn't be a big thing and would make this a working package.

TODOs:
- [ ] Make sure the files in `MANIFEST.in` (and that file itself) are copied to the location where `setup.py install` is run by `make pyinstall`. Suggestion: `pyinstall -> install`
- [ ] Rename to `odl_cuda` or similar?
- [ ] Add reasonable git version tag
- [ ] Bump ODL version so this package can depend on ODL with the plugin interface implemented
- [ ] Add more information to the `setup.py` file (see the one in ODL for a template)
- [ ] Add/update copyright notices
- [ ] Freshen up build instructions (if necessary) and README
- [ ] Write a `bld.bat` for conda Windows builds. This should be possible starting from the `build.sh` for Unix. One can use a similar approach with [environment variables](http://conda.pydata.org/docs/building/environment-vars.html) such that dependencies can be found. Should be done in the `conda_build` branch though.

EDIT: added main TODO to TODO list.

**Comment by adler-j** [comment]
I'll try to get this done once I get back to KTH after midsummer, agree that it should be prio and we need to get it in rather soon. It would also enable the conda branch.

The checklist is very nicely compiled and should cover most things.

Regarding conda however, having a single branch for linux, getting that merged (which seems close?) and then making a new branch for windows would be totally fine.

**Comment by kohr-h** [comment]
> Regarding conda however, having a single branch for linux, getting that merged (which seems close?) and then making a new branch for windows would be totally fine.

Sure.

**Comment by adler-j** [comment]
Any chance you can push your changes to the branch so I can keep working from there?

**Comment by kohr-h** [comment]
If you don't need conda stuff, simply cherry-pick 71db46cd. It contains the MANIFEST.in file. No further relevant changes.

**Pull request merged**